File: packages/antgrid-wire/tests/peer-transport-vectors.test.ts
import { expect, test } from "bun:test";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { buildPeerTransportVectors } from "../scripts/gen-peer-transport-vectors";
import {
  MAX_TRANSFER_BYTES,
  PEER_MAX_BRIDGE_RECORD_BYTES,
  PEER_MAX_RECORD_BYTES,
  PEER_QUIC_KEEP_ALIVE_INTERVAL_MS,
  PEER_QUIC_MAX_IDLE_TIMEOUT_MS,
  STREAM_MAX_UPLOAD_STREAMS_PER_PEER,
  STREAM_PROJECT_APP_RECORD_MAX_BYTES,
  STREAM_PROJECT_BRIDGE_RECORD_MAX_BYTES,
  STREAM_TERMINAL_APP_RECORD_MAX_BYTES,
  STREAM_TERMINAL_BRIDGE_RECORD_MAX_BYTES,
  STREAM_TUNNEL_DATA_MAX_BYTES,
  STREAM_TUNNEL_RECORD_MAX_BYTES,
  STREAM_TUNNEL_REQUEST_BODY_MAX_BYTES,
  STREAM_UPLOAD_BRIDGE_RECORD_MAX_BYTES,
  STREAM_UPLOAD_MAX_FILE_NAME_LENGTH,
  STREAM_UPLOAD_MAX_MIME_TYPE_LENGTH,
  StreamOpen,
  StreamRefused,
  StreamRefusedCode,
  TUNNEL_RECORD_TAG_WS_BINARY,
  TUNNEL_RECORD_TAG_WS_TEXT,
  decodePeerFrame,
} from "../src/index";

const fixturePath = resolve(
  import.meta.dir,
  "../../../evals/fixtures/peer-transport-vectors.json",
);
const fixture = JSON.parse(readFileSync(fixturePath, "utf8"));

test("peer transport fixture is a clean generator product", () => {
  expect(fixture).toEqual(buildPeerTransportVectors());
});

test("peer transport byte vectors decode with no route identity, one sample per header kind", () => {
  expect(fixture.framing.samples.map((s: { name: string }) => s.name)).toEqual(["message", "session"]);
  for (const sample of fixture.framing.samples) {
    const decoded = decodePeerFrame(Buffer.from(sample.frameHex, "hex"));
    expect(decoded.header).toEqual(sample.header);
    expect(decoded.header).not.toHaveProperty("to");
    expect(decoded.header).not.toHaveProperty("from");
    expect(decoded.header).not.toHaveProperty("channel");
    expect(Buffer.from(decoded.payload).toString("hex")).toBe(sample.payloadHex);
  }
  expect(fixture.framing.maxPayloadBytes).toBe(MAX_TRANSFER_BYTES);
  expect(fixture.framing.maxRecordBytes).toBe(PEER_MAX_RECORD_BYTES);
  expect(fixture.framing.maxBridgeRecordBytes).toBe(PEER_MAX_BRIDGE_RECORD_BYTES);
});

test("peer transport fixture covers every stream-open kind and refusal code", () => {
  const streamOpen = fixture.streamOpen;
  expect(streamOpen.opens.map((o: { name: string }) => o.name)).toEqual([
    "session",
    "project",
    "terminal",
    "terminal-with-checkout",
    "tunnel-http",
    "tunnel-ws",
    "upload",
    "upload-with-checkout-and-mime",
  ]);
  for (const sample of streamOpen.opens) {
    expect(StreamOpen.parse(sample.json)).toEqual(sample.json);
  }
  expect(streamOpen.refusals.map((r: { name: string }) => r.name)).toEqual([
    "not-ready",
    "update-required",
    "not-allowed",
    "cap-exceeded",
    "invalid",
  ]);
  for (const sample of streamOpen.refusals) {
    expect(StreamRefused.parse(sample.json)).toEqual(sample.json);
  }
  // Derived from the schema, not a hand list: a kind or code added to
  // stream-open.ts without a vector must fail here, since the fixture is the
  // Dart mirror's only cross-check.
  const kinds = StreamOpen.options.map((o) => o.shape.kind.value);
  expect(
    new Set(streamOpen.opens.map((o: { json: { kind: string } }) => o.json.kind)),
  ).toEqual(new Set(kinds));
  expect(
    streamOpen.refusals.map((r: { json: { code: string } }) => r.json.code),
  ).toEqual(StreamRefusedCode.options);
});

test("peer transport fixture's projectRecords covers both directions plus the transfer cap", () => {
  // The Dart mirror (kStreamProjectAppRecordMaxBytes /
  // kStreamProjectBridgeRecordMaxBytes) has no import to cross-check against,
  // only this JSON.
  expect(fixture.streamOpen.projectRecords).toEqual({
    appMaxRecordBytes: STREAM_PROJECT_APP_RECORD_MAX_BYTES,
    bridgeMaxRecordBytes: STREAM_PROJECT_BRIDGE_RECORD_MAX_BYTES,
    maxTransferBytes: MAX_TRANSFER_BYTES,
  });
});

test("peer transport fixture's terminalRecords and tunnelRecords equal the wire package's own constants", () => {
  // The generator imports these constants directly, so this mostly guards
  // against the fixture going stale relative to a regenerate — the Dart
  // mirror has no import to cross-check against, only this JSON.
  expect(fixture.streamOpen.terminalRecords).toEqual({
    appMaxRecordBytes: STREAM_TERMINAL_APP_RECORD_MAX_BYTES,
    bridgeMaxRecordBytes: STREAM_TERMINAL_BRIDGE_RECORD_MAX_BYTES,
  });
  // Tunnel HTTP bodies travel raw, so only the WS tags exist on the wire.
  expect(fixture.streamOpen.tunnelRecords).toEqual({
    maxDataBytes: STREAM_TUNNEL_DATA_MAX_BYTES,
    maxRecordBytes: STREAM_TUNNEL_RECORD_MAX_BYTES,
    requestBodyMaxBytes: STREAM_TUNNEL_REQUEST_BODY_MAX_BYTES,
    tags: {
      wsText: TUNNEL_RECORD_TAG_WS_TEXT,
      wsBinary: TUNNEL_RECORD_TAG_WS_BINARY,
    },
  });
  expect(fixture.streamOpen.tunnelRecords.tags).not.toHaveProperty("body");
  expect(fixture.streamOpen.tunnelRecords.tags).not.toHaveProperty("bodyGzip");
});

test("peer transport fixture's caps.maxUploadStreamsPerPeer and uploadRecords equal the wire package's own constants", () => {
  expect(fixture.streamOpen.caps.maxUploadStreamsPerPeer).toBe(STREAM_MAX_UPLOAD_STREAMS_PER_PEER);
  expect(fixture.streamOpen.uploadRecords).toEqual({
    bridgeMaxRecordBytes: STREAM_UPLOAD_BRIDGE_RECORD_MAX_BYTES,
    maxFileNameLength: STREAM_UPLOAD_MAX_FILE_NAME_LENGTH,
    maxMimeTypeLength: STREAM_UPLOAD_MAX_MIME_TYPE_LENGTH,
  });
});

test("peer transport fixture's rejectedOpens cover every upload rejection row named in the contract", () => {
  const names = fixture.streamOpen.rejectedOpens.map((o: { name: string }) => o.name);
  for (const expected of [
    "upload-missing-size",
    "upload-negative-size",
    "upload-fractional-size",
    "upload-string-size",
    "upload-empty-file-name",
    "upload-overlong-file-name",
    "upload-null-mime",
    "upload-empty-checkout",
    "upload-extra-field",
  ]) {
    expect(names, `missing rejectedOpens row: ${expected}`).toContain(expected);
  }
});

test("peer transport fixture's streamOpen.labels match every kind's own registry label", () => {
  const labels = fixture.streamOpen.labels as Array<{ name: string; open: unknown; streamKind: string; streamId: string }>;
  const byName = new Map(labels.map((l) => [l.name, l]));
  expect(byName.get("session")).toMatchObject({ streamKind: "session", streamId: "0" });
  expect(byName.get("project")).toMatchObject({ streamKind: "project", streamId: "proj-1" });
  expect(byName.get("terminal")).toMatchObject({ streamKind: "terminal", streamId: "req-1" });
  expect(byName.get("tunnel-http")).toMatchObject({ streamKind: "tunnel-http", streamId: "req-1" });
  expect(byName.get("tunnel-ws")).toMatchObject({ streamKind: "tunnel-ws", streamId: "ws-1" });
  expect(byName.get("upload")).toMatchObject({ streamKind: "upload", streamId: "req-1" });
  // Every label's `open` must itself be a valid StreamOpen — the generator
  // hand-writes these labels (bridge's streamLabelOf is ELv2 and can't be
  // imported here), so this is the only check that they stay in sync with it.
  for (const label of labels) {
    expect(StreamOpen.safeParse(label.open).success, label.name).toBe(true);
  }
});

test("peer transport fixture's rejected stream-open frames are rejected", () => {
  const streamOpen = fixture.streamOpen;
  expect(streamOpen.rejectedOpens.length).toBeGreaterThan(0);
  for (const sample of streamOpen.rejectedOpens) {
    expect(StreamOpen.safeParse(sample.json).success, sample.name).toBe(false);
  }
  expect(streamOpen.rejectedRefusals.length).toBeGreaterThan(0);
  for (const sample of streamOpen.rejectedRefusals) {
    expect(StreamRefused.safeParse(sample.json).success, sample.name).toBe(false);
  }
});

test("fixture's quic block equals PEER_QUIC_KEEP_ALIVE_INTERVAL_MS/PEER_QUIC_MAX_IDLE_TIMEOUT_MS", () => {
  // Neither binding exposes a transport config, so these are iroh 1.0's own
  // defaults, recorded rather than set — this is the Dart mirror's only
  // cross-check, same as every other quic/stream-open constant here.
  expect(fixture.quic).toEqual({
    keepAliveIntervalMs: PEER_QUIC_KEEP_ALIVE_INTERVAL_MS,
    maxIdleTimeoutMs: PEER_QUIC_MAX_IDLE_TIMEOUT_MS,
  });
});

test("the fixture carries no fragmentation or flow-control blocks", () => {
  expect(fixture.fragmentation).toBeUndefined();
  expect(fixture.flowControl).toBeUndefined();
});

File: packages/antgrid_peer_transport/test/peer_transport_vectors_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:antgrid_peer_transport/antgrid_peer_transport.dart';
import 'package:antgrid_relay_client/antgrid_relay_client.dart';
import 'package:test/test.dart';

Map<String, dynamic> _map(Object? value) =>
    Map<String, dynamic>.from(value! as Map);

Uint8List _hex(String value) => Uint8List.fromList([
  for (var i = 0; i < value.length; i += 2)
    int.parse(value.substring(i, i + 2), radix: 16),
]);

String _toHex(Uint8List value) =>
    value.map((byte) => byte.toRadixString(16).padLeft(2, '0')).join();

void main() {
  final fixture = _map(
    jsonDecode(
      File(
        '../../evals/fixtures/peer-transport-vectors.json',
      ).readAsStringSync(),
    ),
  );

  test(
    'Dart framing and native constants match the shared transport vector',
    () {
      final framing = _map(fixture['framing']);
      expect(peerFrameVersion, framing['version']);
      expect(peerFrameFixedPrefix, framing['fixedPrefixBytes']);
      expect(maxPeerFrameHeaderBytes, framing['maxHeaderBytes']);
      expect(kMaxTransferBytes, framing['maxPayloadBytes']);
      expect(kPeerMaxRecordBytes, framing['maxRecordBytes']);
      expect(kPeerMaxBridgeRecordBytes, framing['maxBridgeRecordBytes']);
      expect(FrameKind.message.wireValue, _map(framing['kinds'])['message']);

      final native = _map(fixture['native']);
      expect(peerAlpn, native['alpn']);
      expect(maxPeerLeaseMs, native['leaseMs']);
      expect(maxPeerLeaseMs ~/ 3, native['refreshMs']);
      expect(native['selectionMs'], 5000);
      expect(native['endpointChallengeMs'], 120000);
    },
  );

  test('Dart encodes and decodes every peer-frame golden byte vector', () {
    final samples = (_map(fixture['framing'])['samples'] as List)
        .cast<Map<String, dynamic>>();
    for (final sample in samples) {
      final header = _map(sample['header']);
      final kind = FrameKind.fromWire(sample['kind'] as int)!;
      expect(kind, FrameKind.message, reason: sample['name'] as String);
      final payload = _hex(sample['payloadHex'] as String);
      final encoded = encodePeerFrame(header, payload);
      expect(
        _toHex(encoded),
        sample['frameHex'],
        reason: sample['name'] as String,
      );

      final decoded = decodePeerFrame(_hex(sample['frameHex'] as String));
      expect(decoded.header, header);
      expect(decoded.payload, payload);
      expect(decoded.header.containsKey('to'), isFalse);
      expect(decoded.header.containsKey('from'), isFalse);
    }
  });

  test('Dart authorization bounds match the shared transport vector', () {
    final authorization = _map(fixture['authorization']);
    expect(peerIdentityMaxChars, authorization['identityMaxChars']);
    expect(maxAuthorizedPeers, authorization['maxAuthorizedPeers']);
    expect(maxPeerRelayUrls, authorization['maxRelayUrls']);
    expect(maxPeerGeneration, authorization['maxGeneration']);
    expect(maxPeerLeaseMs, authorization['maxLeaseMs']);
  });

  test('Dart stream-open caps match the shared transport vector', () {
    final streamOpen = _map(fixture['streamOpen']);
    expect(kStreamOpenMaxBytes, streamOpen['maxOpenBytes']);
    expect(kStreamOpenMaxIdLength, streamOpen['maxIdLength']);
    final caps = _map(streamOpen['caps']);
    expect(kStreamMaxBidiStreamsPerConnection, caps['maxBidiStreamsPerConnection']);
    expect(kStreamMaxProjectsPerPeer, caps['maxProjectsPerPeer']);
    expect(kStreamMaxTerminalAttachmentsPerPeer, caps['maxTerminalAttachmentsPerPeer']);
    expect(kStreamMaxTunnelStreamsPerPeer, caps['maxTunnelStreamsPerPeer']);
    expect(kStreamMaxPendingOpensPerPeer, caps['maxPendingOpensPerPeer']);
    expect(kStreamMaxUploadStreamsPerPeer, caps['maxUploadStreamsPerPeer']);
    final uploadRecords = _map(streamOpen['uploadRecords']);
    expect(kStreamUploadBridgeRecordMaxBytes, uploadRecords['bridgeMaxRecordBytes']);
    expect(kStreamUploadMaxFileNameLength, uploadRecords['maxFileNameLength']);
    expect(kStreamUploadMaxMimeTypeLength, uploadRecords['maxMimeTypeLength']);
    final projectRecords = _map(streamOpen['projectRecords']);
    expect(kStreamProjectAppRecordMaxBytes, projectRecords['appMaxRecordBytes']);
    expect(
      kStreamProjectBridgeRecordMaxBytes,
      projectRecords['bridgeMaxRecordBytes'],
    );
    expect(kMaxTransferBytes, projectRecords['maxTransferBytes']);
    final terminalRecords = _map(streamOpen['terminalRecords']);
    expect(kStreamTerminalAppRecordMaxBytes, terminalRecords['appMaxRecordBytes']);
    expect(kStreamTerminalBridgeRecordMaxBytes, terminalRecords['bridgeMaxRecordBytes']);
    final tunnelRecords = _map(streamOpen['tunnelRecords']);
    expect(kStreamTunnelDataMaxBytes, tunnelRecords['maxDataBytes']);
    expect(kStreamTunnelRecordMaxBytes, tunnelRecords['maxRecordBytes']);
    expect(kStreamTunnelRequestBodyMaxBytes, tunnelRecords['requestBodyMaxBytes']);
    final tags = _map(tunnelRecords['tags']);
    expect(kTunnelRecordTagWsText, tags['wsText']);
    expect(kTunnelRecordTagWsBinary, tags['wsBinary']);
    expect(tags.keys, unorderedEquals(['wsText', 'wsBinary']));
  });

  test('Dart parses every stream-open kind golden vector, and round-trips it', () {
    final opens = (_map(fixture['streamOpen'])['opens'] as List)
        .cast<Map<String, dynamic>>();
    expect(opens.map((o) => o['name']), [
      'session',
      'project',
      'terminal',
      'terminal-with-checkout',
      'tunnel-http',
      'tunnel-ws',
      'upload',
      'upload-with-checkout-and-mime',
    ]);
    for (final sample in opens) {
      final json = _map(sample['json']);
      final parsed = StreamOpen.fromJson(json);
      expect(parsed, isNotNull, reason: sample['name'] as String);
      expect(parsed!.toJson(), json, reason: sample['name'] as String);
      switch (sample['name']) {
        case 'session':
          expect(parsed, isA<SessionStreamOpen>());
        case 'project':
          expect(parsed, isA<ProjectStreamOpen>());
        case 'terminal':
        case 'terminal-with-checkout':
          expect(parsed, isA<TerminalStreamOpen>());
        case 'tunnel-http':
          expect(parsed, isA<TunnelHttpStreamOpen>());
        case 'tunnel-ws':
          expect(parsed, isA<TunnelWsStreamOpen>());
        case 'upload':
        case 'upload-with-checkout-and-mime':
          expect(parsed, isA<UploadStreamOpen>());
      }
    }
  });

  test('Dart stream labels match the shared netwatch label vectors', () {
    final labels = (_map(fixture['streamOpen'])['labels'] as List)
        .cast<Map<String, dynamic>>();
    expect(labels, isNotEmpty);
    for (final sample in labels) {
      final open = StreamOpen.fromJson(_map(sample['open']));
      expect(open, isNotNull, reason: sample['name'] as String);
      final label = streamLabelOf(open!);
      expect(label.kind, sample['streamKind'], reason: sample['name'] as String);
      expect(label.id, sample['streamId'], reason: sample['name'] as String);
    }
  });

  test('Dart rejects every stream-open frame the schema rejects', () {
    final rejected = (_map(fixture['streamOpen'])['rejectedOpens'] as List)
        .cast<Map<String, dynamic>>();
    expect(rejected, isNotEmpty);
    for (final sample in rejected) {
      expect(
        StreamOpen.fromJson(_map(sample['json'])),
        isNull,
        reason: sample['name'] as String,
      );
    }
  });

  test('Dart rejects every stream-refused record the schema rejects', () {
    final rejected = (_map(fixture['streamOpen'])['rejectedRefusals'] as List)
        .cast<Map<String, dynamic>>();
    expect(rejected, isNotEmpty);
    for (final sample in rejected) {
      expect(
        StreamRefused.fromJson(_map(sample['json'])),
        isNull,
        reason: sample['name'] as String,
      );
    }
  });

  test('Dart parses every stream-refused code golden vector, and round-trips it', () {
    final refusals = (_map(fixture['streamOpen'])['refusals'] as List)
        .cast<Map<String, dynamic>>();
    expect(refusals.map((r) => r['name']), [
      'not-ready',
      'update-required',
      'not-allowed',
      'cap-exceeded',
      'invalid',
    ]);
    for (final sample in refusals) {
      final json = _map(sample['json']);
      final parsed = StreamRefused.fromJson(json);
      expect(parsed, isNotNull, reason: sample['name'] as String);
      expect(parsed!.toJson(), json, reason: sample['name'] as String);
    }
    // A Dart-only code would pass every other check here.
    expect(
      refusals.map((r) => _map(r['json'])['code']).toList(),
      StreamRefusedCode.values.map((c) => c.wireValue).toList(),
    );
  });

  test('Dart QUIC timing constants match the shared transport vector', () {
    final quic = _map(fixture['quic']);
    expect(kPeerQuicKeepAliveInterval.inMilliseconds, quic['keepAliveIntervalMs']);
    expect(kPeerQuicMaxIdleTimeout.inMilliseconds, quic['maxIdleTimeoutMs']);
  });
}
